File: 2024/day17.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type Registers struct {
	A, B, C int
}

func getComboVal(op int, regs Registers) int {
	switch {
	case op < 4:
		return op
	case op == 4:
		return regs.A
	case op == 5:
		return regs.B
	case op == 6:
		return regs.C
	default:
		return 0
	}
}

func solve(regs Registers, program []int) []int {
	outputs := []int{}
	ip := 0

	for ip < len(program) {
		op := program[ip]
		operand := 0
		if ip+1 < len(program) {
			operand = program[ip+1]
		}

		switch op {
		case 0: // adv
			regs.A /= 1 << getComboVal(operand, regs)
		case 1: // bxl
			regs.B ^= operand
		case 2: // bst
			regs.B = getComboVal(operand, regs) % 8
		case 3: // jnz
			if regs.A != 0 {
				ip = operand
				continue
			}
		case 4: // bxc
			regs.B ^= regs.C
		case 5: // out
			outputs = append(outputs, getComboVal(operand, regs)%8)
		case 6: // bdv
			regs.B = regs.A / (1 << getComboVal(operand, regs))
		case 7: // cdv
			regs.C = regs.A / (1 << getComboVal(operand, regs))
		}
		ip += 2
	}
	return outputs
}

func Day17_1() {
	lines := strings.Split(strings.TrimSpace(string(ReadInput(17, false))), "\n")

	regs := Registers{}
	regs.A, _ = strconv.Atoi(strings.Split(lines[0], ": ")[1])
	regs.B, _ = strconv.Atoi(strings.Split(lines[1], ": ")[1])
	regs.C, _ = strconv.Atoi(strings.Split(lines[2], ": ")[1])

	programStr := strings.Split(lines[len(lines)-1], ": ")[1]
	programNums := strings.Split(programStr, ",")
	program := make([]int, len(programNums))
	for i, num := range programNums {
		program[i], _ = strconv.Atoi(num)
	}

	// Run program and join outputs
	outputs := solve(regs, program)
	result := make([]string, len(outputs))
	for i, v := range outputs {
		result[i] = strconv.Itoa(v)
	}

	fmt.Println(strings.Join(result, ","))
}

File: 2024/day17.py
# opcodes
# 0 - adv - performs division
# 1 - mul - bitwise XOR of B and literal
# 2 - bst value of combo modulo 8
# 3 - jnz does nothing if A is 0
# 4 - bxc - withwise XOR of B and C
# 5 - out - claculates the value of combo modulo 8
# 6 - bdv - adv but with B
# 7 - cdv - adv but with C

# The value of a combo operand can be found as follows:
# Combo operands 0 through 3 represent literal values 0 through 3.
# Combo operand 4 represents the value of register A.
# Combo operand 5 represents the value of register B.
# Combo operand 6 represents the value of register C.
# Combo operand 7 is reserved and will not appear in valid programs.


with open("data/input17.txt", "r") as f:
    data = f.read()


def solve(registers, program):
    def get_combo_val(op):
        if op < 4:
            return op
        elif op == 4:
            return registers["A"]
        elif op == 5:
            return registers["B"]
        elif op == 6:
            return registers["C"]

    outputs = []
    init = 0

    while init < len(program):
        op = program[init]
        operand = program[init + 1] if init + 1 < len(program) else 0

        if op == 0:  # adv - divide A by 2^combo_operand
            registers["A"] //= 2 ** get_combo_val(operand)
        elif op == 1:  # bxl - XOR of B and literal
            registers["B"] ^= operand
        elif op == 2:  # bst - store combo_operand mod 8 in B
            registers["B"] = get_combo_val(operand) % 8
        elif op == 3:  # jnz - jump if A is not zero
            if registers["A"] != 0:
                init = operand
                continue
        elif op == 4:  # bxc - XOR B with C
            registers["B"] ^= registers["C"]
        elif op == 5:  # out - output combo_operand mod 8
            outputs.append(str(get_combo_val(operand) % 8))
        elif op == 6:  # bdv - like adv but store in B
            registers["B"] = registers["A"] // 2 ** get_combo_val(operand)
        elif op == 7:  # cdv - like adv but store in C
            registers["C"] = registers["A"] // 2 ** get_combo_val(operand)

        init += 2

    return outputs

def part1(data):
    lines = data.split("\n")
    registers = {
        "A": int(lines[0].split(": ")[1]),
        "B": int(lines[1].split(": ")[1]),
        "C": int(lines[2].split(": ")[1]),
    }

    program = [int(x) for x in lines[-1].split(": ")[1].split(",")]

    outputs = solve(registers, program)
    return ",".join(outputs)

print(part1(data))
